feature/discard-task-packet-jaia-1989

edsanville opened the pull request:
* Allow user to exclude/include task packets from consideration in the contour/interpolated drifts.
* The excluded task packets get exiled to their own "Excluded Packets" layer

Bonus stuff that I added to make my testing life easier.  These don't have to be part of this PR, the reviewers can decide:
* Persist the visibility of the measurement layers between browser sessions (localStorage)
* Persist the date range of the task packets between browser sessions

Missing:
* I have an "Include" and "Exclude" button, instead of the toggle switch.  I can convert this to a toggle switch, it's just slightly more involved given the way the task packet code is factored.

edsanville committed:
Add get_task_packet_id function to uniquely identify a task packet (second or microsecond precision)

edsanville committed:
task-packets endpoint now returns a JaiaResponse with a result field containing "included" and "excluded" fields.

edsanville committed:
Use task-packets-version instead of task-packets-count to determine if the task packet set has changed

edsanville committed:
Spin off TaskPacketDatabase into its own class

edsanville committed:
all_task_packets is a dict mapping the task_packet_id to the task_packet, instead of a list of task_packets

edsanville committed:
STASHING IN THIS COMMIT, NOT DONE.

edsanville committed:
Persist the excluded task packet ids
Add POST task-packet-include endpoint to exclude/include a task packet

edsanville committed:
Added Include and Exclude buttons to the TaskPacketPanel

edsanville committed:
Add persistance to the visibility of the measurement layers
Fix pre-existing typo (seems to not affect behavior)

edsanville committed:
Persist the taskPacketsTimeline

edsanville committed:
Add excluded task packets layer

edsanville commented on src/web/containers/TaskPacketPanel/TaskPacketPanel.tsx:
```diff
@@ -3,16 +3,61 @@ import Icon from "@mdi/react";
-    delete taskPacketData?.title;
+    const task_packet = props.selectedTaskPacket;
+
+    // Probably want to refactor this into a type with optional name, value, units
```
We can standardize on TODO, so it's easy to search.

edsanville commented on src/web/containers/TaskPacketPanel/TaskPacketPanel.tsx:
```diff
@@ -3,16 +3,61 @@ import Icon from "@mdi/react";
+            start_time: { value: startTime.toLocaleString(), units: "" },
+            end_time: { value: endTime.toLocaleString(), units: "" },
+        };
+    } else if (props.type == "drift") {
```
The refactor I summarized would really be just an interface called maybe `DataRow`.  Quick example:

```
interface DataRow {
  name: string
  value: string
  units: string
}
```

That would make the function a lot more legible I think, especially the JSX.  I'm picturing something like pseudocode:

```
<div>
  rows.map((row) => {
    return <div>
      <div ...>{row.name}</div>
      <div ...>{row.value}</div>
      <div ...>{row.units}</div> 
    </div>
  }
</div>
```

edsanville replied:
To avoid that "drfit" issue though, yeah, it might be advisable to have an enum for that somewhere.

edsanville commented on src/web/containers/TaskPacketPanel/TaskPacketPanel.tsx:
```diff
@@ -3,16 +3,61 @@ import Icon from "@mdi/react";
+        const endTime = new Date(task_packet.end_time / 1000);
+        taskPacketData = {
+            // Snake case used for string parsing in task packet panel
+            bot_id: { value: task_packet.bot_id.toString(), units: "" },
```
Yes, that taskPacketData should absolutely be refactored into an interface.  I wanted to do that, but was afraid of "mission creep."

edsanville commented on src/web/server/utils.py:
```diff
@@ -0,0 +1,58 @@
+    return int(datetime.now().timestamp() * 1e6)
+
+
+def utcnow():
```
Looking at this function, it's actually dead code and never gets called.  Which is probably a good thing, since it returns the wrong answer.  The now_utime function is the one to use to get the true timestamp in microseconds.  I will remove the broken utcnow function.

There are a couple other places in the code where this sort of thing is done:
```
utime = datetime.datetime.utcnow().timestamp()
```

Which seems like it's the thing to do, but it actually doesn't work properly.  This is because `datetime.utcnow()`, for some unknown reason, returns a python "naive datetime" that has no timezone.  So, when calculating the timestamp, it assumes the system timezone, which results in an incorrect timestamp.  I will post to the demerits channel.

Good catch!

edsanville committed:
Add TODO markers
Remove the broken utcnow() function from utils.py

ferrom-jaia commented on src/web/server/app.py:
```diff
@@ -243,11 +243,18 @@ def getPackets():
+    return JaiaResponse(result)
+
+@app.route('/jaia/v0/task-packets-version', methods=['GET'])
 def getPacketsCount():
```
I think this function should get updated to reflect what it is doing.

edsanville committed:
Trigger immediate task packet poll when a task packet is included or excluded.

edsanville committed:
Add progress indicator (most of the time this will be super quick)

edsanville committed:
Rename function to reflect the correct endpoint

edsanville committed:
Merge remote-tracking branch 'origin/1.y' into feature/discard-task-packet-jaia-1989

ferrom-jaia committed:
Merge branch '1.y' into feature/discard-task-packet-jaia-1989

edsanville committed:
Revert "task/3d-depth-plot-set-z-axis-jaia-1990 (#1188)"

Un-persisting the measurement layer visibilities.

This reverts commit 39343b88b9b025dcca9bc5025527220bfdcdb105.

edsanville committed:
Revert "Persist the taskPacketsTimeline"

This reverts commit 2211daff0d525c32ec2975590d13b42b0d9197d9.

edsanville committed:
Immediately create the /var/log/jaiabot/bot_offload/excluded_task_packet_ids.json file if it doesn't exist, to prevent warnings.
Issue INFO instead of WARNING on missing file.

edsanville committed:
Move task_packet_database module into pyjaia, for access from rest_api v1

edsanville committed:
Move some utility functions intp pyjaia.utils (used by task_packet_database)

edsanville committed:
REST API v1 uses the TaskPacketDatabase class to access to offloaded and incoming task packets.

edsanville committed:
Only load modified taskpacket files.

edsanville committed:
Add script to generate a large number of task packets for testing.

edsanville committed:
Use an indexed sqlite database for speed

edsanville committed:
Use threading.Lock to ensure no two python threads accesses the database concurrently
Move taskpacket files into .processed, so we can avoid iterating over them on update
Added timing test to rest api

edsanville committed:
Merge branch '1.y' into feature/discard-task-packet-jaia-1989

edsanville committed:
Move the taskpacket files into a subdirectory, rather than add the .processed extension to the filenames

edsanville committed:
Add a hard limit of 1000 task packets for query to return

ferrom-jaia committed:
Merge branch '1.y' into feature/discard-task-packet-jaia-1989

edsanville committed:
Remove debug logging I accidentally left in

edsanville committed:
Merge branch 'feature/discard-task-packet-jaia-1989' of github.com:jaiarobotics/jaiabot into feature/discard-task-packet-jaia-1989

edsanville committed:
Fix syntax for updating included status of task packet

edsanville committed:
Don't increment the task packet version unless a new taskpacket file was actually processed

edsanville committed:
Don't need to iterate over the features on any particular layer...  just animate the feature that's passed to the function

ferrom-jaia commented on src/python/pyjaia/src/pyjaia/task_packet_database.py:
```diff
@@ -7,47 +7,111 @@
+    
+
+    def _update(self):
+        """Updates taskpacket files from disk, if enough time has passed.  THREAD UNSAFE.
```
I think they are there so if a new developer comes in they know which function to use

edsanville replied:
Yes, that's why.  The user should make sure they only call those functions when the `self.lock` is engaged.  I also used the underscore prefix to make people think twice before calling those functions.

edsanville commented on src/python/pyjaia/src/pyjaia/utils.py:
```diff
@@ -8,3 +10,44 @@ def myip():
+
+    # Combine the bot_id and start_time (rounded to the nearest second)
+    bot_id = task_packet["bot_id"]
+    SECOND = 1_000_000
```
Yeah, it's great for readability!

edsanville commented on src/python/pyjaia/tests/test_task_packet_database.py:
Yes, ideally we would have some good tests for CircleCI.  The two files in this directory were really just ad-hoc scripts to help while coding.  It was probably a bad idea to even call them "tests,"  but I also wanted to get them into the repository to help with dev work.

I can integrate a test framework and put together some real tests for the task packet database and/or the contour generation.  Should it be a separate ticket @ferrom-jaia ?

ferrom-jaia replied:
Yes, I will make a follow ticket for this task.

ferrom-jaia replied:
https://jaia-innovation.atlassian.net/browse/JAIA-1998

edsanville replied:
If anyone has a favorite python test framework, we can use that.  I've got no experience with python test frameworks.

edsanville committed:
Expand the time range by a second on either side to account for dccl rounding of the start_time for task packets

edsanville committed:
Merge branch 'feature/discard-task-packet-jaia-1989' of github.com:jaiarobotics/jaiabot into feature/discard-task-packet-jaia-1989

edsanville committed:
Fix exception when trying to iterate over simplices=None, when Qhul hass difficulties due to degeneracies, etc.

ferrom-jaia reviewed (approved):
Looks good! Tested in sim and everything checked out. Going to test in the field before releasing.
